src/basic/glob-util.c

```diff
@@ -6,81 +6,83 @@
 #include "dirent-util.h"
 #include "errno-util.h"
 #include "glob-util.h"
-#include "log.h"
 #include "string-util.h"
 #include "strv.h"
 
-static void closedir_wrapper(void* v) {
+static void closedir_wrapper(void *v) {
         (void) closedir(v);
 }
 
-int safe_glob(const char *path, int flags, glob_t *pglob) {
-        int k;
+int safe_glob_full(const char *path, int flags, opendir_t opendir_func, char ***ret) {
+        _cleanup_(globfree) glob_t g = {
+                .gl_closedir = closedir_wrapper,
+                .gl_readdir = (struct dirent* (*)(void *)) readdir_no_dot,
+                .gl_opendir = (void* (*)(const char *)) (opendir_func ?: opendir),
+                .gl_lstat = lstat,
+                .gl_stat = stat,
+        };
+        int r;
 
-        /* We want to set GLOB_ALTDIRFUNC ourselves, don't allow it to be set. */
-        assert(!(flags & GLOB_ALTDIRFUNC));
-
-        if (!pglob->gl_closedir)
-                pglob->gl_closedir = closedir_wrapper;
-        if (!pglob->gl_readdir)
-                pglob->gl_readdir = (struct dirent *(*)(void *)) readdir_no_dot;
-        if (!pglob->gl_opendir)
-                pglob->gl_opendir = (void *(*)(const char *)) opendir;
-        if (!pglob->gl_lstat)
-                pglob->gl_lstat = lstat;
-        if (!pglob->gl_stat)
-                pglob->gl_stat = stat;
+        assert(path);
 
         errno = 0;
-        k = glob(path, flags | GLOB_ALTDIRFUNC, NULL, pglob);
-        if (k == GLOB_NOMATCH)
+        r = glob(path, flags | GLOB_ALTDIRFUNC, NULL, &g);
+        if (r == GLOB_NOMATCH)
                 return -ENOENT;
-        if (k == GLOB_NOSPACE)
+        if (r == GLOB_NOSPACE)
                 return -ENOMEM;
-        if (k != 0)
+        if (r != 0)
                 return errno_or_else(EIO);
-        if (strv_isempty(pglob->gl_pathv))
+
+        if (strv_isempty(g.gl_pathv))
                 return -ENOENT;
 
+        if (ret) {
+                *ret = TAKE_PTR(g.gl_pathv);
```

Comment:
TAKE_PTR seems unnecessary? Just add a comment that globfree destructs the resulting strv only?